### ISSUE TITLE
stuck in a #dendrite:matrix.org related loop, heavy resource usage

### ISSUE THREAD

**Issue by bones-was-here** [matrix-org/dendrite#2181]
### Background information

- **Dendrite version or git SHA**: 0.6.3 and 5106cc807cf22a95420b24f6bfdd5c9ac8aa06de
- **Monolith or Polylith?**: monolith
- **SQLite3 or Postgres?**: postgres
- **Running in Docker?**: no
- **`go version`**: 1.17.6
- **Client used (if applicable)**: element, curl

### Description

 - **What** is the problem: dendrite is stuck in a loop printing the same errors
 - **Who** is affected: 
 - **How** is this bug manifesting: steadily increasing ram and jetstream storage size
 - **When** did this first appear: some time after updating to 0.6.3

### Steps to reproduce

```
level=error msg="syncapi: failed to QuerySharedUsers for key change event from key server" error="found 2577 users but only have state key nids for 2350 of them"
level=error msg="syncapi: failed to QuerySharedUsers for key change event from key server" error="found 2448 users but only have state key nids for 2221 of them"
```
- it has something to do with #dendrite:matrix.org because that's the only room with that many users that this server has participated in
- these print at about 8 lines per second continuously
- restarting and deleting ~/jetstream have no effect
- ~/jetstream is growing steadily
- joining and leaving #dendrite:matrix.org has no effect
- forgetting the room has no effect #2176 

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from server logs, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.

Alternatively, please send logs to @kegan:matrix.org or @neilalexander:matrix.org
with a link to the respective Github issue, thanks!
-->

**Comment by bones-was-here** [comment]
Eventually it panics, restarts, and continues the loop

**Comment by bones-was-here** [comment]
It eventually stopped after many hours and several GB of jetstream dir.  Now I can restart it and delete jetstream and it continues behaving.  Have not attempted to rejoin #dendrite:matrix.org :)

**Comment by neilalexander** [comment]
Can you please run this query against your roomserver database?
```
SELECT COUNT(target_nid) FROM roomserver_membership AS m WHERE NOT EXISTS (
	SELECT event_state_key_nid FROM roomserver_event_state_keys AS s
	WHERE m.target_nid = s.event_state_key_nid
);
```

**Comment by bones-was-here** [comment]
This dendrite is still running 5106cc807cf22a95420b24f6bfdd5c9ac8aa06de and is not currently looping those errors at time of query.

**Comment by bones-was-here** [comment]
Currently it's printing  
`level=error msg="syncapi: failed to QuerySharedUsers for key change event from key server" error="found 2623 users but only have state key nids for 2395 of them"`  
about once per second and continuing after restart.  
Slower than the last time, and every error is the same whereas previously it alternated between two slightly different ones.  This is 002429c9e24cc746e0929b41eccbe429f89a6e1f built with golang 1.17.7.

**Comment by bones-was-here** [comment]
Should I delete this? `4.2G    jetstream/$G/streams/DendriteOutputKeyChangeEvent`

**Comment by neilalexander** [comment]
Should be OK to.

**Comment by alistair23** [comment]
Seems ok at first, I am having issues with new sessions though, I opened: https://github.com/matrix-org/dendrite/issues/2222

**Comment by bones-was-here** [comment]
Much better on CPU so far, even when it's obviously working on something.  

It still has "slow" federation for #dendrite.xonotic.org with occasional groups of messages from weeks or months ago appearing.

Tends to log `level=warning msg="SelectJoinedUsersSetForRooms found 2632 users but BulkSelectEventStateKey only returned state key NIDs for 2404 of them"` sometimes 1-2 per second for ages.

**Comment by bones-was-here** [comment]
I spoke too soon, the CPU burn started again, not sure why, it seems inconsistent.  

Updating to master to get 4c07374c42e5d671cfb137634475f43f84f9db0e seems to have given a significant reduction in peak memory and average CPU load during the burn, but has not fixed the problems with this room.  It's still a significant load considering how little work the homeserver _should_ be doing.  

Possibly it gets worse when people talk in #dendrite:matrix.org, and there's still specific users whose messages never appear.  

`level=warning msg="SelectJoinedUsersSetForRooms found 2648 users but BulkSelectEventStateKey only returned state key NIDs for 2421 of them"` continues to print at about two per second while the CPU burn is happening.

**Comment by bones-was-here** [comment]
[dendrite_profile_20220226_1.gz](https://github.com/matrix-org/dendrite/files/8143304/dendrite_profile_20220226_1.gz)

**Comment by bones-was-here** [comment]
The heavy CPU load was caused by clients spamming and was fixed by #2233.  
#2234 and #2237 fixed the state key NIDs warnings.
